Augment all error messages in the numerics/ directory

bangerth opened the pull request:
This is again in partial fulfillment of #610.

tjhei commented on include/deal.II/numerics/data_out_stack.h:
```diff
@@ -304,7 +314,8 @@ class DataOutStack : public DataOutInterface<dim+1>
   DeclException1 (ExcInvalidNumberOfSubdivisions,
```
This and some of the other exceptions are duplicated exactly. I vote for either moving them to a generic header or declaring them once and using a typedef in the other places.

tjhei commented on include/deal.II/numerics/point_value_history.h:
```diff
@@ -521,36 +521,40 @@ class PointValueHistory
    */
-  DeclException0(ExcDoFHandlerChanged);
+  DeclExceptionMsg(ExcDoFHandlerChanged,
+                   "The triangulation indicated that mesh has been refined in some way. This "
```
weird grammar?

tjhei commented on include/deal.II/numerics/solution_transfer.h:
```diff
@@ -373,37 +373,28 @@ class SolutionTransfer
-  DeclException0(ExcNumberOfDoFsPerCellHasChanged);
+  DeclExceptionMsg(ExcAlreadyPrepForCoarseAndRef,
+                   "You are attempting to call one of the prepare_*() functions "
+                   "of this object to prepare it for an operation for which  it "
```
Use only one space. Maybe remove this exception and replace with ExcAlreadyPrepForRef?

tjhei commented on include/deal.II/numerics/solution_transfer.h:
```diff
@@ -373,37 +373,28 @@ class SolutionTransfer
-  DeclException0(ExcAlreadyPrepForRef);
+  DeclExceptionMsg(ExcAlreadyPrepForRef,
+                   "You are attempting to call one of the prepare_*() functions "
+                   "of this object to prepare it for an operation for which  it "
```
use only one space

tjhei commented on include/deal.II/numerics/point_value_history.h:
```diff
@@ -521,36 +521,40 @@ class PointValueHistory
-  DeclException0(ExcDataLostSync);
+  DeclExceptionMsg(ExcDataLostSync,
+                   "This error is thrown to indicate that the data sets appear to be out of "
+                   "sync.  The class requires that the number of dataset keys is the same as "
```
Use only one space.

tjhei commented on include/deal.II/numerics/solution_transfer.h:
```diff
@@ -373,37 +373,28 @@ class SolutionTransfer
+  DeclExceptionMsg(ExcAlreadyPrepForCoarseAndRef,
+                   "You are attempting to call one of the prepare_*() functions "
+                   "of this object to prepare it for an operation for which  it "
+                   "it is already prepared.");
```
"it it"?

tjhei commented on include/deal.II/numerics/data_out_rotation.h:
```diff
@@ -184,13 +184,20 @@ class DataOutRotation : public DataOut_DoFData<DH,DH::dimension+1>
                   int,
                   << "The number of subdivisions per patch, " << arg1
-                  << ", is not valid.");
+                  << ", is not valid. It needs to be greater or equal to "
```
This is kind of confusing because 0 is a perfectly fine default for ```n_patches``` in ```build_patches``` etc..

bangerth commented:
I've addressed all of your comments, @tjhei . 

The last of these commits unifies exceptions in a namespace of its own. There is no magic here, just a lot of code churn.

tjhei commented on include/deal.II/numerics/data_out_dof_data.h:
```diff
@@ -37,6 +37,109 @@ DEAL_II_NAMESPACE_OPEN
+                  int, int,
+                  << "You have to give one name per component in your "
+                  << "data vector. The number you gave was " << arg1
+                  << ", but the number of components is " << arg2);
```
missing period at the end.

tjhei commented on include/deal.II/numerics/data_out_dof_data.h:
```diff
@@ -37,6 +37,109 @@ DEAL_II_NAMESPACE_OPEN
+                  << "vectors must always consist of exactly <dim>\n"
+                  << "components. However, the vector component at\n"
+                  << "position " << arg1 << " with name <" << arg2
+                  << "> does not satisfy these conditions.");
```
why do you have all these newlines here?

tjhei commented:
Thanks for doing all the cleanup. Looks much better than before. I especially like that exceptions are no longer referenced via things like ```typename DofData<dim,dim>::SomeException```.

tamiko commented:
ack.

bangerth commented:
I've addressed Timo's points (no idea where the '\n' came from -- I just copy-pasted).

I've also taken the liberty to rename the new namespace `Exceptions::DataOut` -- we may acquire more of these namespace `Exceptions::*` over time.

tjhei commented:
> I've also taken the liberty to rename the new namespace Exceptions::DataOut -- we may acquire more of these namespace Exceptions::* over time.

Good idea.
